## Homework

The goal of this homework is to train a simple model for predicting the duration of a ride - similar to what we did in this module.


In [1]:
import pandas as pd
import sklearn
import numpy as np

from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

from sklearn.metrics import mean_squared_error

## Q1. Downloading the data

We'll use [the same NYC taxi dataset](https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page),
but instead of "**Green** Taxi Trip Records", we'll use "**Yellow** Taxi Trip Records".

Download the data for January and February 2023.

Read the data for January. How many columns are there?

* 16
* 17
* 18
* 19 <-- correct


In [ ]:
jan = pd.read_parquet('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet ')
# feb = pd.read_parquet('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet ')

In [3]:
# count number of columns in jan
len(jan.columns)

19

In [5]:
jan.head(3)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3    1.0      0.5         0.0           0.0   
1             1          7.9    1.0      0.5         4.0           0.0   
2             1         14.9    1.0      0.5        15.0           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0          14.3                   2.5          0.0  
1                    1.0          16.9                   2.5          0.0  
2                    1.0          34.9                   2.5          0.0

## Q2. Computing duration

Now let's compute the `duration` variable. It should contain the duration of a ride in minutes. 

What's the standard deviation of the trips duration in January?

* 32.59
* 42.59 <-- 
* 52.59
* 62.59

In [6]:
# compute duration, convert to minutes
jan['duration'] = (jan['tpep_dropoff_datetime'] - jan['tpep_pickup_datetime']).dt.total_seconds() / 60

In [7]:
# standard deviation of the trips duration in Jan
jan['duration'].std()

42.594351241920904

## Q3. Dropping outliers

Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

* 90%
* 92%
* 95%
* 98% <--

In [8]:
print('original size: ', jan.shape[0])
# remove rows with duration outside 1 and 60 minutes
jan = jan[(jan['duration'] >= 1) & (jan['duration'] <= 60)]

# get the fraction of records left after dropping
print('after dropping outliers: ', jan.shape[0])

original size:  3066766
after dropping outliers:  3009173



## Q4. One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model. 

* Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will 
  label encode them)
* Fit a dictionary vectorizer 
* Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

* 2
* 155
* 345
* 515 <-->
* 715

In [ ]:
def read_dataframe(filename):
    if filename.endswith('.parquet'):
        df = pd.read_parquet(filename)

    # compute duration, convert to minutes
    df['duration'] = (df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60

    # remove outliers
    df = df[(df.duration >= 1) & (df.duration <= 60)]

    # convert pickup/dropoff location IDs to strings so DictVectorizer treats them as categorical
    categorical = ['PULocationID', 'DOLocationID']
    df[categorical] = df[categorical].astype(str)
    
    return df

In [ ]:
# load training and validation data
df_train = read_dataframe('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet') # jan data
df_val = read_dataframe('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-02.parquet') # feb data

In [ ]:
# set up feature columns
categorical = ['PULocationID', 'DOLocationID']
numerical = ['trip_distance']

# initialize the DictVectorizer to one-hot encode categorical features from a list of dictionaries
dv = DictVectorizer()

# convert the categorical features into a list of dictionaries
train_dicts = df_train[categorical].to_dict(orient='records')
# fit the vectorizer to the training data (learn the unique categories),
# and transform the dict into a sparse feature matrix (rows = trips, columns = one hot encoded categories))
X_train = dv.fit_transform(train_dicts)

# print dimension of the feature matrix
print(f'feature matrix shape: {X_train.shape}')

feature matrix shape: (3009173, 515)


Notes on `dv.fit_transform(train_dicts)` and `dv.transform(val_dicts)`:

- fit_transform: does 2 things at once
    - fit:
        - learn the vocabulary of features, i.e. all values of PULocationID and DOLocationID
        - builds the mapping: {
  'PULocationID=132': 0,
  'DOLocationID=237': 1,
  ...}
    - transform:
        - converts each dictionary to a sparse one-hot encoded vector using the learned mapping
- transform:
    - only transforms using the learned mapping, does not learn new categories
    - If a PULocationID or DOLocationID in the validation data was seen in training, it will be properly one-hot encoded.
    - If it's unseen, it will be ignored (encoded as all zeros for that feature).


In [ ]:
# do the same converting to dict and then one hot encoding for validation data
val_dicts = df_val[categorical].to_dict(orient='records')
# use the already fitted DictVectorizer to transform ths val data to the same feature space
X_val = dv.transform(val_dicts)

In [ ]:
# prepare target variable
target = 'duration'
# extract target variable as numpy array
y_train = df_train[target].values
y_val = df_val[target].values

In [ ]:
# train the linear regression model
lr = LinearRegression()
lr.fit(X_train, y_train)

# predict on the training data
y_pred = lr.predict(X_train)

# compute RMSE
np.sqrt(mean_squared_error(y_train, y_pred))

7.649261927686161

In [ ]:
# repeats model training and prediction on the validation set
lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred = lr.predict(X_val)

# compute RMSE
np.sqrt(mean_squared_error(y_val, y_pred))

7.811817957524739


## Q5. Training a model

Now let's use the feature matrix from the previous step to train a model. 

* Train a plain linear regression model with default parameters, where duration is the response variable
* Calculate the RMSE of the model on the training data

What's the RMSE on train?

* 3.64
* 7.64 <>
* 11.64
* 16.64


## Q6. Evaluating the model

Now let's apply this model to the validation dataset (February 2023). 

What's the RMSE on validation?

* 3.81
* 7.81 <>
* 11.81
* 16.81



## Submit the results

* Submit your results here: https://courses.datatalks.club/mlops-zoomcamp-2025/homework/hw1
* If your answer doesn't match options exactly, select the closest one